src/main/java/org/codehaus/plexus/components/secdispatcher/PinEntry.java

```diff
@@ -0,0 +1,229 @@
+        Process process = new ProcessBuilder(cmd).start();
+        BufferedReader reader = process.inputReader();
+        BufferedWriter writer = process.outputWriter();
+        expectOK(process.inputReader());
+        Map.Entry<String, String> lastEntry = commands.entrySet().iterator().next();
+        for (Map.Entry<String, String> entry : commands.entrySet()) {
+            String cmd;
+            if (entry.getValue() != null) {
+                cmd = entry.getKey() + " " + entry.getValue();
+            } else {
+                cmd = entry.getKey();
+            }
+            logger.debug("> {}", cmd);
+            writer.write(cmd);
+            writer.newLine();
+            writer.flush();
+            if (entry != lastEntry) {
+                expectOK(reader);
+            }
+        }
+        Result<String> result = lastExpect(reader);
+        writer.write("BYE");
+        writer.newLine();
+        writer.flush();
+        try {
+            process.waitFor(5, TimeUnit.SECONDS);
+            int exitCode = process.exitValue();
+            if (exitCode != 0) {
+                return new Result<>(Outcome.FAILED, "Exit code: " + exitCode);
+            } else {
+                return result;
+            }
+        } catch (Exception e) {
+            return new Result<>(Outcome.FAILED, e.getMessage());
+        }
+    }
+
+    private void expectOK(BufferedReader in) throws IOException {
+        String response = in.readLine();
+        logger.debug("< {}", response);
+        if (!response.startsWith("OK")) {
+            throw new IOException("Expected OK but got this instead: " + response);
+        }
+    }
+
+    private Result<String> lastExpect(BufferedReader in) throws IOException {
+        while (true) {
+            String response = in.readLine();
+            logger.debug("< {}", response);
+            if (response.startsWith("#")) {
+                continue;
+            }
+            if (response.startsWith("S")) {
+                continue;
+            }
+            if (response.startsWith("ERR")) {
+                if (response.contains("83886142")) {
+                    return new Result<>(Outcome.TIMEOUT, response);
+                }
+                if (response.contains("83886179")) {
+                    return new Result<>(Outcome.CANCELED, response);
+                }
+                if (response.contains("83886194")) {
+                    return new Result<>(Outcome.NOT_CONFIRMED, response);
+                }
+            }
+            if (response.startsWith("D")) {
+                return new Result<>(Outcome.SUCCESS, response.substring(2));
+            }
+            if (response.startsWith("OK")) {
+                return new Result<>(Outcome.SUCCESS, response);
+            }
+        }
+    }
+
+    public static void main(String[] args) throws IOException {
+        Result<String> pinResult = new PinEntry("/usr/bin/pinentry-gnome3")
```

Comment:
on macos we have:
```
% which pinentry                                          
/opt/homebrew/bin/pinentry

% which pinentry-mac 
/opt/homebrew/bin/pinentry-mac

```

Reply:
Yes, change the executable and just run main from IDE to see it in action :smile:

Reply:
My original goal was to make it work with https://github.com/jorgelbg/pinentry-touchid (so maven encryption with touchID on mac, yay!) but the project seems dead or unsure, but the app did not work on my mac with macOS 15.0.1

Reply:
Also, this is the important bit:
```
.setKeyInfo("maven:masterPassword")
```
as this is the "handle" for sensitive data, so this is the name of your master pw in keychain, if it is in keychain. According to pinentry doco, this is usually either keyGrip (for PGP keys) or "app:something" formed id.

Reply:
So, this is how my `settings-security.xml` looks like:
```xml
<?xml version='1.0' encoding='UTF-8'?>
<settingsSecurity xmlns="http://codehaus-plexus.github.io/plexus-sec-dispatcher/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://codehaus-plexus.github.io/plexus-sec-dispatcher/4.0.0 https://codehaus-plexus.github.io/xsd/plexus-sec-dispatcher-4.0.0.xsd">
  <modelVersion>4.0</modelVersion>
  <defaultDispatcher>master</defaultDispatcher>
  <configurations>
    <configuration>
      <name>master</name>
      <properties>
        <property>
          <name>source</name>
          <value>pinentry-prompt:/usr/bin/pinentry-gnome3</value>
        </property>
        <property>
          <name>cipher</name>
          <value>AES/GCM/NoPadding</value>
        </property>
      </properties>
    </configuration>
  </configurations>
</settingsSecurity>
```
As you guessed, format is `pinentry-prompt:$pathToPinentry`. Just make yours like this, or use `mvnenc init` and then edit it manually (as mvnenc is not yet fully done).

Reply:
No, I do not. mvn3 encryption is _so badly broken_, I can share the details in private, but I have no intent to make it backward compat.

Reply:
Instead, we may want to upgrade sec-dispatcher to 4.0 in mvn 3.10? But we would lose "mvnenc init" (use mvn4 for that), and also `--encrypt-master-password` as well...

Reply:
> Do you think about add a backward compatibility ...

Oh, now I realized: mvn3 would fail to load the new format... so maybe we need new file name / path for sec-dispatcher 4?

Reply:
Radically simplifies this library, removed all mumbo jumbo about default config file, and Java System property overrides, it is now purely the duty of integrating app how will this be integrated. Also, integrating app "knows" what it needs, and using globals like Java system properties is anyway something is not nice from library.

Reply:
Backward compat solved @slawekjaranowski Now mvn4 will warn if legacy encrypted pw found, and am adding new goal "migrate". If you use mvn3 and mvn4 on same workstation and you insist on encryption, you must use the (broken) mvn3 encryption, and live with warnings mvn4 will emit.